src/DataGrid.tsx

```diff
@@ -332,13 +418,114 @@ function DataGrid<R, K extends keyof R, SR>({
     onColumnResize?.(column.idx, width);
   }, [columnWidths, onColumnResize]);
 
-  function handleRowsUpdate(event: RowsUpdateEvent) {
-    onRowsUpdate?.(event);
+  function handleCommit({ cellKey, rowIdx, updated }: CommitEvent) {
+    onRowsUpdate?.({
+      cellKey,
+      fromRow: rowIdx,
+      toRow: rowIdx,
+      updated,
+      action: UpdateActions.CELL_UPDATE
+    });
+
+    setSelectedPosition(({ idx, rowIdx }) => ({ idx, rowIdx, mode: 'SELECT' }));
+  }
+
+  function handleDragEnd() {
+    if (typeof draggedOverRowIdx === 'undefined') return;
+
+    const { idx, rowIdx } = selectedPosition;
+    const column = columns[idx];
+    const cellKey = column.key;
+    const value = rows[rowIdx][cellKey as keyof R];
+
+    onRowsUpdate?.({
+      cellKey,
+      fromRow: rowIdx,
+      toRow: draggedOverRowIdx,
+      updated: { [cellKey]: value } as unknown as never,
+      action: UpdateActions.CELL_DRAG
+    });
+
+    setDraggedOverRowIdx(undefined);
+  }
+
+  function handleDragHandleDoubleClick(): void {
+    const column = columns[selectedPosition.idx];
+    const cellKey = column.key;
+    const value = rows[selectedPosition.rowIdx][cellKey as keyof R];
+
+    onRowsUpdate?.({
+      cellKey,
+      fromRow: selectedPosition.rowIdx,
+      toRow: rows.length - 1,
+      updated: { [cellKey]: value } as unknown as never,
+      action: UpdateActions.COLUMN_FILL
+    });
+  }
+
+  function handleCopy() {
+    const { idx, rowIdx } = selectedPosition;
+    const value = rows[rowIdx][columns[idx].key as keyof R];
+    setCopiedPosition({ idx, rowIdx, value });
+  }
+
+  function handlePaste() {
+    if (copiedPosition === null || !isCellEditable(selectedPosition)) {
```

Comment:
Maybe also check if the `copiedPosition` is same as the `selectCellPosition` to skip dispatching the update event?